pkg/controller/backupstoragelocation_controller.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2020 the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/vmware-tanzu/velero/internal/velero"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+)
+
+// BackupStorageLocationReconciler reconciles a BackupStorageLocation object
+type BackupStorageLocationReconciler struct {
+	Scheme          *runtime.Scheme
+	StorageLocation velero.StorageLocation
+
+	Log logrus.FieldLogger
+}
+
+// +kubebuilder:rbac:groups=velero.io,resources=backupstoragelocations,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=velero.io,resources=backupstoragelocations/status,verbs=get;update;patch
+func (r *BackupStorageLocationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithField("controller", "backupstoragelocation")
+
+	log.Info("Checking for existing backup locations ready to be verified; there needs to be at least 1 backup location available")
+
+	locationList, err := velero.ListBackupStorageLocations(r.StorageLocation.Client, r.StorageLocation.Ctx, req.Namespace)
+	if err != nil {
+		log.WithError(err).Error("No backup storage locations found, at least one is required")
```

Comment:
Should we just return the error?
Currently it doesn't matter because we loop over the `locationList.Items` and when there is an error that list is empty. But, returning the error makes the current code easier to read and also any logic we add in the future after line 51 can assume that there are items to reconcile and not run into seg faults or perform nil checks.

Reply:
I was trying to avoid having to duplicate `r.logReconciledPhase(defaultFound, locationList, unavailableErrors)` here as well or, worse, the inconsistency of having that call when there are items but not when there are none.

Reply:
If we return the error, how does that affect the reconciliation loop? Does it get re-invoked?

Reply:
Yes, it can make it get requeued.

Reply:
Ok, I think the way this is working now I'm good with it. We can refine things over the course of the release cycle, like potentially adding a back off period via [`Result.RequeueAfter`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.5.2/pkg/reconcile?tab=doc#Result) if none are found, but let's do that in another PR if we really need it.